---
title: "Spectral Estimates of Inbreeding Coefficients for Binary and Surrogate Genotypes"
author: "Clement Gain - Olivier  Francois"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Spectral Estimates of Inbreeding Coefficients for Binary and Surrogate Genotypes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



**Summary** : The R package `spectralfst` implements spectral methods to evaluate population structure and to compute Wright's inbreeding coefficients for binary (classical) and surrogate genotypes. These methods are based on a the connection between Fst and the eigenvalues of principal component analysis (PCA), which is the most-frequently used approach to describe population genetic structure from large population genomic data sets. In this vignette, we illustrate the analysis of population structure and residual variation for a three population model. Then, we evaluate adjusted Fst values after correction of the genotype matrix for variation associated with bioclimatic variables. The adjusted Fst values represent the amount of inbreeding explained by divergence dissociated from the studied environmental conditions. 



```{r}
library(spectralfst)
```

## Principle of the methods

Let $Z$ be genotype matrix for $n$ haploid or diploid individuals genotyped at $L$ loci. The individuals are sampled from $K$ discrete populations. To establish relationships between PCA and inbreeding coefficients, we decomposed the centered genotype matrix as a sum of two matrices  
$$Z = Z_{ST}  + Z_{S} \, , $$

where $Z_{ST}$ is a **between-population** matrix describing the variation of allelic frequencies among populations, and

+ $Z_{S}$ is **within-population matrix**, describing residual genetic variation not explained by population subdivision,

+ both matrices are divided by $\sqrt{n}$.


For $K \geq 2$ populations, the main result of the theory can be stated as follows. Let $Z_{ST}^{sc}$ be the scaled version of $Z_{ST}$, where each column is divided by their standard error. We have 
$$E[F_{ST}] = \sum_{k=1}^{K-1}{\rho _{k}^{2}(Z_{ST}^{sc}) / L} \, ,$$

where $\rho _{k}^2(Z_{ST}^{sc})$ are the first $K-1$ eigenvalues of the empirical correlation matrix of $Z_{ST}^{sc}$ (computed by scaled PCA). In addition, if we assume that the residual matrix is random, *i.e.*, has its eigenvalues described by random matrix theory, then the average $Fst$ corresponds to the sum of scaled PCA eigenvalues for the genotype matrix ($Z$)

$$E[F_{ST}] \approx \sum_{k=1}^{K-1}{\rho _{k}^{2}(Z^{sc}) / L} \, .$$

In other words, the mean value of $F_{ST}$ across loci can be approximated from the sum of the $(K −1)$ leading eigenvalues of the PCA of scaled genotype matrix. The reference at the end of the vignette provides information on the accuracy of this approximation.



## Spectral theory for 3 populations models


This section illustrates the use of the partitioning of the genotype matrix (`compute_partition` function) for simulated genotypes in a three-population model.

### Three populations models

We simulated an $F$-model for $K=3$ discrete populations having diverged from an ancestral gene pool. In the ancestral gene pool, the derived allele was present with frequency $p_{anc}$ (beta(1,4) distribution). The populations diverged from each other and from the ancestral population with respective drift coefficients equal to $F_{1} = 0.3$, $F_{2} = 0.2$ and $F_{3} = 0.1$ relative to the ancestral pool. Fifty haploid individuals were sampled from each population.

Simulations of $F$-model could be performed in `spectralfst` as follows
```{r, include=FALSE}
obj <-  fmodel_simulation(n=150, L=20000, c_vector=c(1/3,1/3,1/3), F_vector=c(0.3,0.2,0.1))
```


For the results to be fully reproductible, we used an example from the library. The genotype matrix contains $n = 150$ individuals genotyped at $L = 18967$ loci.


```{r}
data('fmodel')
X <- fmodel$genotype
pop <- fmodel$pop
dim(X)
```




### Average value of Wright’s $F_{ST}$ accross loci using Wright's formula

At a given locus, Wright’s $F_{ST}$ can be computed by using the following formula
$$ \frac{PQ - \sum_{k=1}^{K}c_{k}p_{k}q_{k}}{PQ} \, .$$

where P is the allele frequency in the total population, Q = 1 - P, $c_{k}$ is the proportion of individuals sampled from population $k$, $p_{k}$ is the allele frequency in population $k$, and $q_{k}$ = $1 - p_{k}$. The function `mean_wright_fst` computes the average value of Wright's inbreeding coefficients across the loci included in the genotype matrix.

```{r}
# Mean fst from Wright's formula
fst_wright <- mean_wright_fst(X, pop)
fst_wright
```

For the three-population example, the mean value of Fst is equal to 0.1318689. 

### Spectral estimates of Wright’s Fst across loci

Now we apply the `compute_partition` function to the genotype matrix. The function returns the scaled between-population matrix, $Z_{ST}$, and the residual matrix, $Z_{S}$ for the population considered. The objective of the example is to illustrate the theory and show that the approximation of Fst from PCA eigenvalues is accurate.  

First, we compute the scaled between-population, $Z_{ST}^{sc}$, and check that summing the first and second eigenvalues leads to the exact same estimate of  $F_{ST}$ as previously.

```{r}
#Compute Fst from the between-population matrix Zst
spectral_fst <- compute_partition(X, pop)
spectral_fst$Fst
```

As expected, the spectral estimate is exactly equal to the mean value of $F_{ST}$ (0.1318689). Next, we can check that the leading eigenvalue of the residual matrix separates from the second eigenvalue of the between-population matrix.

```{r}
#Leading eigenvalue of Zs/sqrt(n)
spectral_fst$leadingeigenZs
```

 Random matrix theory predicts that the leading eigenvalue of $Z_{S}/\sqrt{n}$ is of order $(1-F_{\rm ST})(1/\sqrt{n-K} + 1/\sqrt{L})^2$. Here the prediction is accurate

```{r}
#RMT prediction of the leading eigenvalue of Zs/sqrt(n)
round(spectral_fst$RMTprediction, digits = 4)
```


```{r}
#Smallest non-null eigenvalue of Zst/sqrt(n) (= second one)
spectral_fst$eigenZst[2]
```

The leading eigenvalue of the residual matrix is substantially smaller than the second eigenvalue of the between-population matrix. Under those conditions, the value of $F_{ST}$ across loci is expected to be close to the sum of the (K −1) leading eigenvalues of the PCA. We can check if it is indeed the case in the simulated model.

```{r}
# Fst approximation using PCA
spectral_fst$Fst_approximation
```

The $F_{ST}$ approximation obtained from the PCA is equal to 0.131951. It is very close to the Wright estimate (0.1318689).

### PCA scree plots and PC plots

PCA provides visual representations for the scaled matrix $Z^{sc}$, for the between-population matrix $Z_{ST}^{sc}$, and for the residual matrix $Z_{S}^{sc}$. The PCA scree plots and PC plots can be analyzed in order to check that the number of population was correct in the partioning of genetic variation.

+ First, let us display the scree plot and the PC plot for the genotype matrix ($Z^{sc}$). 

```{r, fig.show='hold'}
plot(spectral_fst$eigenZ[1:10] * 100, ylab="variance (%)", main="Scree plot", ylim=c(0,8))
plot(spectral_fst$pcZ, col = pop, main="PC Plot")
```

+ Then let us display the scree plot and the PC plot for the between-population matrix ($Z_{ST}^{sc}$). The scree plot and the PC plot are very similar to those obtained for the full matrix.

```{r, fig.show='hold'}
plot(spectral_fst$eigenZst[1:10] * 100, ylab="variance (%)", main="Scree plot", ylim=c(0,8))
spectral_fst$pcZst[,2] <- - spectral_fst$pcZst[,2] #PC signs are arbitrary
plot(spectral_fst$pcZst, col = pop, main="PC Plot")
```



+ Now, let us check residual variation by displaying the scree plot and the PC plot for the within-population matrix ($Z_{S}^{sc}$). The scree plot correspond to the eigenvalues of the full matrix after the elbow, and the PC plot shows that all population structure was captured by a model with three discrete populations. 

```{r, fig.show='hold'}
plot(spectral_fst$eigenZs[1:10] * 100, ylab="variance (%)", main="Scree plot", ylim=c(0,8))
plot(spectral_fst$pcZs, col = pop, main="PC Plot")
```

Checking for residual structure in the  matrix $Z_{S}^{sc}$ is an informal test for population specification of population labels as we shall see in the next part.

### Spectral analysis with incorrectly labelled population samples

For the same simulated genotype matrix, we now change our sampling design. Assuming that there are $K=2$ populations, populations 1 and 2 are grouped against the least diverged population 3. This artificial "paraphyletic" grouping can be detected by spectral analysis of between and within-population matrices. 

```{r}
pop <- c(rep(1,100), rep(2,50))
```

Firstn, the identity between Wright's formula and  the leading eigenvalues of the between-population matrix is always valid. 

```{r}
#Wright's Fst
fst_wright <- mean_wright_fst(X, pop)
fst_wright
```



```{r}
#Squared norm of the between population matrix Zst
spectral_fst <- compute_partition(X, pop)
spectral_fst$Fst
```

However, the leading value of the residual matrix, $Z_{S}^{sc}/\sqrt{n}$, does not verify the separation condition.


```{r}
#Leading eigenvalue of Zs/sqrt(n)
spectral_fst$leadingeigenZs
```

```{r}
#Smallest non zero eigenvalue of Zst/sqrt(n) (first one = Fst)
spectral_fst$eigenZst[1]
spectral_fst$eigenZst[1] > spectral_fst$leadingeigenZs
```


The leading value of the residual matrix als differs from its RMT prediction. As a result, the average of $F_{ST}$ fails to approximate the first eigenvalue of the PCA.



```{r}
#RMT prediction of the leading eigenvalue of Zs/sqrt(n)
(1 - 0.05412755)*(1/sqrt(148) + 1/sqrt(18967))^2
spectral_fst$RMTprediction
```

```{r}
#Fst approximation using PCA of the full matrix
spectral_fst$Fst_approximation
```

The PC plot of the residual matrix $Z_{s}$ clearly exhibits residual structure.

```{r, fig.show='hold'}
plot(spectral_fst$eigenZs[1:10] * 100, ylab="variance (%)", main="Scree plot", ylim=c(0,8))
plot(spectral_fst$pcZs, col = pop, main="PC Plot")
```



## Fst adjusted for genetic variation associated with environmental variables

One main interest of the spectral theory is the straightforward extension to adjusted genotypic matrices, providing statistics analogous to $F_{ST}$ for modified genotypic values. This section illustrates the computation of an adjusted (neutral) Fst, correcting for genetic variation associated with bioclimatic variables in real genotypes from Scandinavian *Arabidopsis thaliana*. 

```{r}
data('athaliana')
```

The `athaliana` object consists of three attributes : 

+ *genotype* : A genotype matrix for 241 (considered haploid) individuals and 10 000 loci, corresponding to 241 swedish plant accessions from The 1,001 Genomes database for *Arabidopsis thaliana*. The 10,000 SNPs were randomly selected from a large set of variants. The matrix contains no missing genotypes.

+ *pop* : A vector of population labels (length = 241). The $i$th element corresponds to the population label of individual $i$. The  individuals  were  clustered  in  two groups  based  on  analysis of population structure and correspond to Southern vs Northern populations in Sweden.

+ *bio* : A matrix of 241 rows and 18 columns. Eighteen  bioclimatic  variables,  derived from the monthly temperature and rainfall values,  were considered as representing the current environmental matrix. Global climate and weather data corresponding to individual geographic coordinates were downloaded from the WorldClim  database  (https://worldclim.org).   

### Fst without correction

First, we compute Wright's $F_{ST}$ (without adjustment) with two methods leading to equal estimates.

```{r}
spectral_athaliana <- compute_partition(athaliana$genotype, athaliana$pop)
wright_fst_athaliana <- mean_wright_fst(athaliana$genotype, athaliana$pop)

#mean Fst from Wright's formula
wright_fst_athaliana

#Fst estimate from the between population matrix Zst
spectral_athaliana$Fst
```

Next we check that the separation condition is verified.

```{r}
#Smallest non zero eigenvalue of Zst
spectral_athaliana$eigenZst[1]
#Leading eigenvalue of Zs
spectral_athaliana$leadingeigenZs
spectral_athaliana$eigenZst[1] > spectral_athaliana$leadingeigenZs
```

 However RMT failed to predict the leading eigenvalue of Zs indicating that there is residual structure contained in $Z_{S}$.  $F_{st}$ (0.079) is not accurately approximated by PCA (0.086) eigenvalues.
 
```{r}
#RMT prediction of the leading eigenvalue of Zs
spectral_athaliana$RMTprediction

#Fst approximation using PCA of the full matrix
spectral_athaliana$Fst_approximation
```




### Adjusted Fst with correction for variation associated with climate


To use the `adjusting_variables` parameter for the matrix of centered genotypes, Z, and the matrix of bioclimatic variables, Y, the program estimates a matrix of surrogate genotypes, W, by adjusting a factor regression model of the form 
$$
{\bf Z} = {\bf Y} {\bf B}^T +  {\bf W} + \epsilon \, .
$$  

This step is performed internally in the `compute_partition` function. Adjusted $F_{st}$ is obtained by applying a spectral estimate on the matrix $Z^{adj} = W + \epsilon$. Adjusted $F_{st}$ can be thought of as a kind of 'neutral' $F_{st}$ obtained after removing all adaptive genetic variation associated with the considered set of environmental variables.

```{r}
spectral_adjusted_athaliana <-compute_partition(athaliana$genotype, athaliana$pop, Y = athaliana$bio)
```

```{r}
# Adjusted Fst 
spectral_adjusted_athaliana$Fst
```

The unadjusted value was equal to 0.0795342. Around 32\% of divergence between populations of A. thaliana two along the South-North axis could be explained by differences in climatic variation. 

```{r}
(spectral_athaliana$Fst - spectral_adjusted_athaliana$Fst)/spectral_athaliana$Fst
```

The scree plot below describes the proportion of variance explained by the first axis of the between-population matrix (Fst and Adjusted Fst), and by the axes of the residual matrix (five components) before adjustment for environmental variables (blue color) and after adjustment (orange color). 

```{r, fig.width = 7, fig.height = 7}
m <- cbind(c(spectral_athaliana$eigenZst[1], spectral_athaliana$eigenZs[1:5]),
           c(spectral_adjusted_athaliana$eigenZst[1], spectral_adjusted_athaliana$eigenZs[1:5]))
plot(seq(1,12)/2,as.numeric(t(m)), 
     col = rep(c("darkblue","orange"), length = 12), 
     type = "h", lwd = 12, 
     ylab = "Proportion of variance",
     xlab = "Axis",
    # main = "Structure Components",
     cex.axis = 2,
    cex = 2,
     las = 1)

legend(x = 3.5, y = 0.077, 
       legend = c("All SNPs", "Adjusted data"),
       col = c("darkblue","orange"),
       lty = c(1,1),
       pch = 19,
       cex = 2)
```



The separation condition is verified, but the eigenvalues are close to each other, revealing that additional hierarchical population structure is present in the data. 

```{r}
#Smallest non zero eigenvalue of Zst
spectral_adjusted_athaliana$eigenZst[1]
#leading eigenvalue of Zs
spectral_adjusted_athaliana$leadingeigenZs
spectral_adjusted_athaliana$eigenZst[1] > spectral_adjusted_athaliana$leadingeigenZs
```




### Reference

Francois O, Gain C (2021) "A Spectral Theory for Wright's Inbreeding Coefficients and Related Quantities". BioRxiv doi: 10.1101/2020.10.07.329755.
